### ISSUE TITLE
Parameters that aren't there: default value vs. UNSET vs. AttributeError

### ISSUE THREAD

**Issue by rossjjennings** [nanograv/PINT#642]
Currently, writing PINT code for a situation where a model parameter may or may not exist is unnecessarily difficult, because the absence of the parameter might be handled in one of several different ways: it might be absent from the model object, so that asking for it raises an `AttributeError`, present, but with the value `UNSET`, so that asking for it as a `value` or `quantity` gives `None`, or present with a default value (usually 0). We should try to standardize this as much as possible. 

I would propose getting rid of `UNSET` entirely, so that looking for a parameter that hasn't been set always raises `AttributeError` unless there is a sensible default value. This might break enough existing code that it would be a pain to implement, though.

**Comment by luojing1211** [comment]
Raise AttributeError will make the current code broken. The reason for a default value parameter is that some model that has some parameters we general does not set, but it is used in the calculation. For instance, in DD model, we do not use DR or DTH very often, but they are used in the calculation with zero value.

**Comment by rossjjennings** [comment]
A particularly annoying instance of this that I ran into is `A1DOT`: If a `DD` binary model doesn't have an `A1DOT`, it is treated as 0, but if an `ELL1` model doesn't have an `A1DOT`, it is `UNSET`.

**Comment by luojing1211** [comment]
It happens here. DD model sets the A1DOT in the model. But ELL1 may not.  But I don't think it is the problem of default value or unset. It is the convention about how we use them is not clear. https://github.com/nanograv/PINT/blob/0f6a1c933de85c235a6d6bd273b0b81492739db3/src/pint/models/binary_dd.py#L80

**Comment by rossjjennings** [comment]
@luojing1211 I agree that it's nice to have some parameters be 0 by default. It's less nice when it only works for some parameters, and you have to sift through the code to figure out which. And having to remember which are `UNSET` and which will raise `AttributeError` is just annoying.

**Comment by luojing1211** [comment]
By the way, where does this AttributeError coming from? Parameters are not added to the timing model?

**Comment by rossjjennings** [comment]
It happens a lot for prefix parameters. For example, if `PBDOT` is not present, it will be unset (so that `model.PBDOT.value is None`), but if `FB1` is not present, it will raise an `AttributeError`. I think there are other cases, too.

**Comment by luojing1211** [comment]
PBDOT should get default value 0. So the set up is PBX and FBX are from two classes for orbit, thus they can not show up together. If one uses PB, FBx would not be there.

**Comment by rossjjennings** [comment]
It makes sense that PB/PBDOT and FB0, FB1, etc. can't be used together. The part that gets annoying is having to figure out that you need to write code like this:
```
if model.PBDOT.quantity is not None:
    PBDOT = model.PBDOT.quantity
    # do stuff that relies on PBDOT
else:
    # PBDOT is absent
```
in the first case, and code like this:
```
try:
    FB1 = model.FB1.quantity
    # do stuff that relies on FB1
except AttributeError:
    # FB1 is absent
```
in the second case.

**Comment by luojing1211** [comment]
Oh, I see. This problem is deeper. FB1 is a prefix parameter, because, one doesn't know how many FBs are there. That is why FB1 may not be there. However, PBDOT is added to the model when initiating.

**Comment by rossjjennings** [comment]
It can also get worse in the `UNSET` case. For example, if you're writing a function that needs to be able to operate on arbitrary model objects (regardless of what components they have), and has to do something special when `A1DOT` is present, you need to write something like this:
```
try:
    A1DOT = model.A1DOT.quantity
    if A1DOT is not None:
        # do stuff that relies on A1DOT
    else:
        # This is a model that can have A1DOT, but A1DOT is unset
except AttributeError:
    # This is a model that can't have A1DOT (e.g. an isolated pulsar model)
```

**Comment by luojing1211** [comment]
Ok, I see what you mean. I agree that reducing from three status to two status is nice.

**Comment by luojing1211** [comment]
@rossjjennings I just merged validate() functions. Maybe we can use validate() to solve the problem of AttributeError?

**Comment by rossjjennings** [comment]
@luojing1211 Maybe? I'm not really sure how it relates.

**Comment by luojing1211** [comment]
Do you have any real use case for this problem? It will help me fix it.

**Comment by rossjjennings** [comment]
This came up in trying to add the `change_binary_epoch()` function. I'm almost done with that -- I'll have the pull request up pretty soon.
